workshop/content/20-typescript/30-hello-cdk/300-cdk-watch.md

```diff
@@ -0,0 +1,177 @@
+{{% /notice %}}
+
+We can speed up that deployment time with `cdk watch`, which watches your
+CDK application and looks for hotswappable changes. Which `cdk watch` finds a
+change that can be hotswapped, it bypasses the CloudFormation deployment entirely
+and instead makes the necessary SDK call to update your code for you. Changes to 
+your AWS Lambda asset code, AWS Step Function State Machine definition, and other
+non-resource changes have the potential to be hotswapped. Here, we will try
+`cdk watch` with your AWS Lambda asset code.
+
+## Modify your `cdk.json` file
+
+When the `cdk watch` command runs, the files that it observes is determined by the
+`"watch"` setting in the `cdk.json` file. It has two sub-keys, `"include"` and
+`"exclude"`, each of which can be either a single string or an array of strings.
+Each entry is interpreted as a path relative to the location of the `cdk.json` file. Globs, both `*` and `**`, are allowed to be used.
+
+Your `cdk.json` file should look like this:
+
+```json
+{
+  "app": "npx ts-node --prefer-ts-exts bin/workshop.ts",
+  "watch": {
+    "include": [
+      "**"
+    ],
+    "exclude": [
+      "README.md",
+      "cdk*.json",
+      "**/*.d.ts",
+      "**/*.js",
+      "tsconfig.json",
+      "package*.json",
+      "yarn.lock",
+      "node_modules",
+      "test"
+    ]
+  },
+  "context": {
+    ...
+  }
+}
+```
+
+As you can see, the sample app comes with a suggested `"watch"` setting. We do in
+fact want to observe our `.js` files in the `lambda` folder, so let's remove
+`"**/*.js"` from the `"exclude"` list:
+
+```json
+{
+  "app": "npx ts-node --prefer-ts-exts bin/workshop.ts",
+  "watch": {
+    "include": [
+      "**"
+    ],
+    "exclude": [
+      "README.md",
+      "cdk*.json",
+      "**/*.d.ts",
+      "tsconfig.json",
+      "package*.json",
+      "yarn.lock",
+      "node_modules",
+      "test"
+    ]
+  },
+  "context": {
+    ...
+  }
+}
+```
+
+Now you're all set to start watching!
+
+## See how long 'CDK Watch' takes
+
+First, call `cdk watch`: 
+
+```bash
+cdk watch
+```
+
+This will trigger an initial deployment and immediately begin observing the files
+you've specified in `cdk.json`.
+
+For the purposes of this exercise, make sure that the initial deployment is
+finished before proceeding. Otherwise, the changes you make will be queued for
+deployment and released _after_ the first deployment finishes.
+
+Let's change our lambda asset code in `lambda/hello.js` one more time. It doesn't
+matter what you change it to, but let's say we want it to say
+`"Good Afternoon, CDK"` now.
+
+Once you click save or hit `CTRL+s`, `cdk watch` will recognize that your file has
+changed and trigger a new deployment. In this case, it will recognize that we can
+hotswap the lambda asset code, so it will bypass a CloudFormation deployment and
+deploy directly to the Lambda service instead.
+
+How fast did deployment take?
```

Comment:
Haha.. can some one actually suggest a good way to time an actual deployment. I saw @corymhall use `esbuild`, is that the general way to go?

Reply:
I actually thought it would be awesome if we printed out, after each deployment in `cdk watch` mode, the total time the deployment took. Like we do now for the `cdk deploy` command, which ends in something like:

```
✨  Done in 78.80s.
```

Would implementing that be something you might be interested in @kaizen3031593? 🙂

Reply:
So `cdk deploy` doesn't actually do this I don't think. It is something that's probably useful though.

Reply:
[It definitely does](https://www.youtube.com/watch?v=XBfgvXEaUz0&t=119s) 🙂.

Reply:
😂 i was literally watching that just now and going crazy. `yarn cdk deploy` spits out the total time afterwards. but `cdk deploy` does not. I confirmed that behavior on my CDK app but can you explain why that's the case?

Reply:
![Screen Shot 2021-12-29 at 3 16 04 PM](https://user-images.githubusercontent.com/36202692/147700055-0179cc59-e7e0-4f07-a328-da1dde8a26c3.png)

Reply:
Oh, it's probably a Yarn feature then, and not a feature of the CDK CLI 😃, like I thought it was. Good catch!

Then yeah, let's add it to `cdk deploy` / `cdk watch`. I'll make an issue for it.

Reply:
https://github.com/aws/aws-cdk/issues/18213